### PR TITLE
[Feature] Add cuda support for Sparse Matrix multiplication, summation and masking

### PR DESCRIPTION
**Pull request by isratnisa**
## Description
This is a follow-up of the CPU functions in #2753 .

Currently int64 sparse matrices are cast into int32 sparse matrices for using cuSPARSE routines.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

**Comment by BarclayII** [comment]
We currently support CUDA 9.2, 10.1, 10.2, 11.0 and 11.1.  Are the cuSPARSE routines very different from each other?

**Comment by BarclayII** [comment]
This PR should be ready for review.

**Comment by BarclayII** [comment]
Need to convert to graph operator first.  Putting reviews on hold.

**Comment by BarclayII** [comment]
Changed to graph interface.  Should be ready for review again.

PyTorch/MXNet/Tensorflow autograd operators as well as user interfaces will be in a separate PR.

File: src/array/cuda/csr_get_data.cu [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    CSRMatrix csr, NDArray rows, NDArray cols, NDArray weights, float filler);
+template NDArray CSRGetData<kDLGPU, int32_t, double>(
+    CSRMatrix csr, NDArray rows, NDArray cols, NDArray weights, double filler);
+template NDArray CSRGetData<kDLGPU, int64_t, double>(
+    CSRMatrix csr, NDArray rows, NDArray cols, NDArray weights, double filler);
+
+template <DLDeviceType XPU, typename IdType>
+NDArray CSRGetData(CSRMatrix csr, NDArray rows, NDArray cols) {
+  return CSRGetData<XPU, IdType, IdType>(csr, rows, cols, NullArray(rows->dtype), -1);
+}
```
Comment (jermainewang) [inline src/array/cuda/csr_get_data.cu:63]: It seems that this implementation is duplicated in `csr_get_data.cc` and `csr_get_data.cu`. Consider putting it directly in `src/array/array_op.h`.

File: src/array/kernel.cc [hunk 2]
```diff
@@ -296,61 +271,81 @@ DGL_REGISTER_GLOBAL("sparse._CAPI_DGLKernelGetEdgeMapping")
...
-        A_data,
-        CSRMatrix(M, N, B_indptr, B_indices));
+    const HeteroGraphRef A_ref = args[0];
+    NDArray A_weights = args[1];
+    const HeteroGraphRef B_ref = args[2];
+
+    const HeteroGraphPtr A = A_ref.sptr();
+    const HeteroGraphPtr B = B_ref.sptr();
+    CHECK_EQ(A->NumEdgeTypes(), 1) << "Graph must have only one edge type.";
+    CHECK_EQ(B->NumEdgeTypes(), 1) << "Graph must have only one edge type.";
```
Comment (jermainewang) [inline src/array/kernel.cc:342]: Also check A and B have the same (1) number of nodes, (2) idtype and (3) device context.

File: src/array/cuda/csr_sum.cu [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    dC_weights};
+}
+}  // namespace cusparse
+
+template <int XPU, typename IdType, typename DType>
+std::pair<CSRMatrix, NDArray> CSRSum(
+    const std::vector<CSRMatrix>& As,
+    const std::vector<NDArray>& A_weights) {
+  CHECK(As.size() > 0) << "List of matrices can't be empty.";
+  CHECK_EQ(As.size(), A_weights.size()) << "List of matrices and weights must have same length";
```
Comment (jermainewang) [inline src/array/cuda/csr_sum.cu:110]: Consider moving these checks to device-agnostic function (e.g., in `src/array/kernel.cc`).

File: src/array/cuda/dispatcher.cuh [hunk 4]
```diff
@@ -0,0 +1,131 @@
+/*!
+ *  Copyright (c) 2020 by Contributors
+ * \file array/cuda/dispatcher.cuh
+ * \brief Dispatches to CUDA functions by data types.
```
Comment (jermainewang) [inline src/array/cuda/dispatcher.cuh:4]: Is this header for compatibility of different cusparse versions? If so, please say it explicitly (I would recommend change the file name as well). Right now, it looks like a header file for general dispatchers.

**Review by jermainewang — CHANGES_REQUESTED** [review]

File: python/dgl/sparse.py [hunk 5]
```diff
@@ -366,124 +366,75 @@ def _bwd_segment_cmp(feat, arg, m):
...
+    A : HeteroGraphIndex
+        The input graph index as left operand.
+    A_weights : Tensor
+        The edge weights of graph A as 1D tensor.
+    B : HeteroGraphIndex
+        The input graph index as right operand.
+    B_weights : Tensor
+        The edge weights of graph B as 1D tensor.
+    num_vtypes : int
+        The number of node types for the returned graph (must be either 1 or 2).
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/sparse.py:384]: No we can't.  For instance multiplying a graph with metagraph `(A, AB, B)` and a graph with `(B, BA, A)` should give the resulting graph `(A, ..., A)` with one node type.

**Comment by jermainewang** [comment]
@dgl-bot run g4dn.4xlarge all dmlc/master

File: python/dgl/sparse.py [hunk 6]
```diff
@@ -366,124 +366,75 @@ def _bwd_segment_cmp(feat, arg, m):
...
-
-def csrsum(As):
-    """Sparse-sparse matrix summation.
-
-    This is an internal function whose interface is subject to changes.
+    C, C_weights = _CAPI_DGLCSRMM(
+        A, F.to_dgl_nd(A_weights), B, F.to_dgl_nd(B_weights), num_vtypes)
+    return C, F.from_dgl_nd(C_weights)
+
+def csrsum(As, A_weights):
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/sparse.py:397]: > Does it mean that the graph must be stored in a CSR format? What if the format is restricted to COO?
It must be in CSR.  cuSPARSE does not support other formats.
> I think "graph_union" is actually a better name.
I don't think so.  This case is more restricted because all graphs must have the same set of nodes.

This function is also for internal use and the user interface will be "graph_sum" or something.

File: python/dgl/sparse.py [hunk 7]
```diff
@@ -366,124 +366,75 @@ def _bwd_segment_cmp(feat, arg, m):
...
-
-    This is an internal function whose interface is subject to changes.
+    C, C_weights = _CAPI_DGLCSRSum(As, [F.to_dgl_nd(w) for w in A_weights])
+    return C, F.from_dgl_nd(C_weights)
+
+def csrmask(A, A_weights, B):
+    """Return the weights of A at the locations identical to the sparsity pattern
+    of B.
+
+    In scipy notation this is identical to ``A[B != 0]``.
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/sparse.py:422]: > In scipy, A[B != 0] does not return a matrix of the same shape of A and B.

And indeed this function returns an array instead of a matrix.

This function together with `csrsum` and `csrmm` are internal functions.

File: src/array/cpu/csr_get_data.cc [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+  const IdType* indptr_data = static_cast<IdType*>(csr.indptr->data);
+  const IdType* indices_data = static_cast<IdType*>(csr.indices->data);
+  const IdType* data = CSRHasData(csr)? static_cast<IdType*>(csr.data->data) : nullptr;
+
+  const int64_t retlen = std::max(rowlen, collen);
+  bool return_eids = IsNullArray(weights);
+  const DType* weight_data = return_eids ? nullptr : weights.Ptr<DType>();
+
+  NDArray ret = Full(filler, retlen, rows->ctx);
+  DType* ret_data = ret.Ptr<DType>();
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply src/array/cpu/csr_get_data.cc:63]: Yes.

File: src/array/cpu/csr_get_data.cc [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+template NDArray CSRGetData<kDLCPU, int32_t, float>(
+    CSRMatrix csr, NDArray rows, NDArray cols, NDArray weights, float filler);
+template NDArray CSRGetData<kDLCPU, int64_t, float>(
+    CSRMatrix csr, NDArray rows, NDArray cols, NDArray weights, float filler);
+template NDArray CSRGetData<kDLCPU, int32_t, double>(
+    CSRMatrix csr, NDArray rows, NDArray cols, NDArray weights, double filler);
+template NDArray CSRGetData<kDLCPU, int64_t, double>(
+    CSRMatrix csr, NDArray rows, NDArray cols, NDArray weights, double filler);
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply src/array/cpu/csr_get_data.cc:111]: That should be fine because we never explicitly call `CSRGetData<kDLCPU, int32_t, int32_t>` in other source code units.

File: src/array/cuda/dispatcher.cuh (hunk 4 above)
Comment (BarclayII) [reply src/array/cuda/dispatcher.cuh:4]: No it's a dispatcher for different input types.  It dispatches float inputs to cusparseScsrgemm2 and double inputs to cusparseDcsrgemm2 for instance.

**Comment by VoVAllen** [comment]
@dgl-bot run g4dn.4xlarge all dmlc/master

**Comment by VoVAllen** [comment]
@dgl-bot run g4dn.4xlarge all dmlc/master

**Comment by VoVAllen** [comment]
@dgl-bot run g4dn.4xlarge all dmlc/master

**Comment by VoVAllen** [comment]
@dgl-bot run g4dn.4xlarge all dmlc/master

**Comment by VoVAllen** [comment]
@dgl-bot run g4dn.4xlarge all dmlc/master

**Comment by VoVAllen** [comment]
@dgl-bot run g4dn.4xlarge all dmlc/master

**Comment by VoVAllen** [comment]
@dgl-bot run g4dn.4xlarge all dmlc/master

**Comment by VoVAllen** [comment]
@dgl-bot run g4dn.4xlarge all dmlc/master

**Comment by VoVAllen** [comment]
@dgl-bot run g4dn.4xlarge all dmlc/master

**Comment by VoVAllen** [comment]
@dgl-bot run g4dn.4xlarge all dmlc/master

**Comment by VoVAllen** [comment]
@dgl-bot run g4dn.4xlarge all dmlc/master

File: include/dgl/kernel.h [hunk 10]
```diff
@@ -73,7 +73,7 @@ std::pair<CSRMatrix, NDArray> CSRSum(
 /*!
  * \brief Return a sparse matrix with the values of A but nonzero entry locations of B.
  */
-NDArray CSRMask(const CSRMatrix& A, NDArray A_weights, const CSRMatrix& B);
+NDArray CSRMask(const CSRMatrix& A, NDArray A_weights, const COOMatrix& B);
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply include/dgl/kernel.h:76]: This function is no longer needed.  I'll remove it.

File: src/array/kernel.cc [hunk 11]
```diff
@@ -296,61 +271,81 @@ DGL_REGISTER_GLOBAL("sparse._CAPI_DGLKernelGetEdgeMapping")
...
+    const HeteroGraphRef A_ref = args[0];
+    NDArray A_weights = args[1];
+    const HeteroGraphRef B_ref = args[2];
+    NDArray B_weights = args[3];
+    int num_vtypes = args[4];
+
+    const HeteroGraphPtr A = A_ref.sptr();
+    const HeteroGraphPtr B = B_ref.sptr();
+    CHECK_EQ(A->NumEdgeTypes(), 1) << "Graph must have one edge type.";
+    CHECK_EQ(A->NumEdgeTypes(), 1) << "Graph must have one edge type.";
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply src/array/kernel.cc:295]: The checks are in CSRMM and CSRSum functions.  Should I move them here?

File: src/array/kernel.cc [hunk 12]
```diff
@@ -296,61 +271,81 @@ DGL_REGISTER_GLOBAL("sparse._CAPI_DGLKernelGetEdgeMapping")
...
-    for (int i = 0; i < A_indptr.size(); ++i)
-      mats[i] = CSRMatrix(M, N, A_indptr[i]->data, A_indices[i]->data);
+    List<HeteroGraphRef> A_refs = args[0];
+    List<Value> A_weights = args[1];
+
+    std::vector<NDArray> weights = ListValueToVector<NDArray>(A_weights);
+    std::vector<CSRMatrix> mats;
+    mats.reserve(A_refs.size());
+    int num_vtypes = 0;
+    for (auto A_ref : A_refs) {
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply src/array/kernel.cc:318]: Again in the function above.

File: src/array/kernel.cc (hunk 2 above)
Comment (BarclayII) [reply src/array/kernel.cc:342]: IDType and context are already checked in `src/array/array.cc`.

**Comment by VoVAllen** [comment]
@dgl-bot run g4dn.4xlarge all dmlc/master

**Comment by VoVAllen** [comment]
@dgl-bot run g4dn.4xlarge all dmlc/master

File: src/array/cpu/csr_get_data.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (BarclayII) [reply src/array/cpu/csr_get_data.cc:111]: Added explicit instantiation since the implementation of `CSRGetData<XPU, IdType>()` is moved to the header file.

File: python/dgl/sparse.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jermainewang) [reply python/dgl/sparse.py:400]: > It must be in CSR. cuSPARSE does not support other formats.

But the argument type is `GraphIndex`, which causes confusion.

File: python/dgl/sparse.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jermainewang) [reply python/dgl/sparse.py:387]: Can we infer the number of node types based on the input without the need of user specification?

File: python/dgl/sparse.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jermainewang) [reply python/dgl/sparse.py:434]: But you should not use "identical to". "similar to" would be better as there is still discrepancy.

File: src/array/kernel.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (jermainewang) [reply src/array/kernel.cc]: I c. Then it's ok.

File: src/array/cpu/csr_get_data.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (jermainewang) [reply src/array/cpu/csr_get_data.cc:63]: Could consider adding a check to avoid memory corruption bug.

File: src/array/cuda/dispatcher.cuh (hunk 4 above)
Comment (jermainewang) [reply src/array/cuda/dispatcher.cuh:4]: I understand. I would name it `cusparse_dispatcher.h` and put down description like "Templates to call different CUSPARSE routines based on the type argument" or similar sorts.

**Comment by VoVAllen** [comment]
@dgl-bot run g4dn.4xlarge all dmlc/master

**Comment by VoVAllen** [comment]
@dgl-bot run g4dn.4xlarge all dmlc/master

**Comment by VoVAllen** [comment]
@dgl-bot run g4dn.4xlarge all dmlc/master

**Comment by VoVAllen** [comment]
@dgl-bot run g4dn.4xlarge all dmlc/master

File: python/dgl/sparse.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/sparse.py:387]: We can infer the number of node types from the metagraphs, but only in the Python because we need to inspect the node type names.  This function is just a bare Python wrapper of the C function so I'm not doing it here.  For the corresponding user interface we won't require the user to specify the number of node types.

File: python/dgl/sparse.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/sparse.py:400]: Then how about we explicitly require the users to enable CSRs in the user interface?  I don't think we want to write COO kernels.

File: python/dgl/sparse.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/sparse.py:434]: They are indeed identical.
* `A[B != 0]` returns an array, and here `csrmask` indeed returns an array.
* `A[B != 0]` will return zeros for the non-existent entries in A, and here `csrmask` also does that.

Or do you think `csrmask` *sounds* like masking an matrix and therefore should return a matrix?  I can change the function name if you want to.

File: src/array/cpu/csr_get_data.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (BarclayII) [reply src/array/cpu/csr_get_data.cc:66]: Done.

File: src/array/cuda/dispatcher.cuh (hunk 4 above)
Comment (BarclayII) [reply src/array/cuda/dispatcher.cuh]: Done.

File: python/dgl/sparse.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jermainewang) [reply python/dgl/sparse.py:400]: Perhaps it's better to discuss this in the PR for user-facing APIs. Let's just take a note here.

**Review by jermainewang — APPROVED** [review]

**Comment by VoVAllen** [comment]
@dgl-bot run g4dn.4xlarge all dmlc/master

**Comment by VoVAllen** [comment]
@dgl-bot run g4dn.4xlarge all dmlc/master

**Comment by VoVAllen** [comment]
Start the Regression test. View at https://ci.dgl.ai/job/dgl-regression/job/PR-2782/2/display/redirect

**Comment by VoVAllen** [comment]
Finished the Regression test. Result table is at https://dgl-asv-data.s3-us-west-2.amazonaws.com/409fd137330db01141b3cc2138c07a54daf3952b_g4dn4xlarge/results/result.csv. Jenkins job link is https://ci.dgl.ai/job/dgl-regression/job/PR-2782/2/display/redirect.
